daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dsom.walker
+
+/**
+ * Below is a series of "View" Mixins defined such that particular simple attributes
+ * of the core DSOM classes are exposed as a Walk is performed.
+ *
+ * Several of them have a walkDSOM method.  By default, this method is called as the walk is performed,
+ * but a user implementing AbstractDSOMWalker can override this behavior by overriding walkFromRoot
+ * and implementing their own event handlers for each of the DSOM types.
+ */
+
+trait CommonContextView {
+  def namespaces: scala.xml.NamespaceBinding
+}
+
+trait TermView extends CommonContextView {
+  def isArray: Boolean
+  def isOptional: Boolean
+}
+
+trait TypeView
+trait SimpleTypeView extends TypeView {
+  def primType: PrimTypeView
+}
+trait ComplexTypeView extends TypeView {
+  def group: ModelGroupView
+}
+
+trait ModelGroupView extends TermView {
+  def groupMembers: Seq[TermView]
+}
+
+trait ChoiceView extends ModelGroupView
+trait SequenceView extends ModelGroupView
+trait GroupRefView extends ModelGroupView {
+  def isHidden: Boolean
+}
+
+trait ElementDeclView extends CommonContextView {
+  def isSimpleType: Boolean
+  def isComplexType: Boolean
+  def simpleType: SimpleTypeView
+  def complexType: ComplexTypeView
+  def hasDefaultValue: Boolean
```

Comment:
Do we need to provide a method to get the default value? I assume if something wants to know about default values it will also want to know what that default it?

Reply:
maybe not? Why do we need to know if something has a default value even? Isn't that a format issue? Maybe we need it to point out an error where a required field can be absent from the NiFi record when unparsing it? But I don't actually see why we need the value.

Reply:
NiFi doesn't need this; I merely included it because I was trying to mimic the structure of `ElementDeclMixin`.  I notice that `ElementDeclMixin` has `defaultValueAsString` member, would that be the correct member to add?  If so, then I could add it, but I could also just remove `hasDefaultValue` from the ElementDeclView if this is not an important attribute.

Reply:
If NiFi doesn't need it, I'd vote for just removing it. We can always add it in the future if some other walker implementation needs it.

Reply:
If f NiFi doesn't need it, definitely leave it off. It's pretty good evidence that perhaps nobody will need it, and every thing we put on these views will cause someone to wonder whether or not they *need to be using it*.